---
title: "Lake 2016 snRNAseq Report"
author:
  - name: "Emir Turkes [et2628@cumc.columbia.edu]"
  - name: "Columbia University"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: "../lake-2016-snRNAseq.bib"
biblio-style: apalike
link-citations: true
output:
  html_document:
    number_sections: true
    theme: lumen
    highlight: zenburn
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit:
  (function(inputFile, encoding) {
    rmarkdown::render(
      inputFile, encoding = encoding, output_file = "../results/lake-2016-snRNAseq-report.html")})
---

```{r, include = FALSE}
#    This file is part of lake-2016-snRNAseq.
#    Copyright (C) 2019  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

knitr::opts_chunk$set(fig.width = 8.5, fig.height = 7, error = TRUE)
```

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This is a broad initial analysis that prepares and characterizes the data for use in other projects.*

The background for this data is as follows:

- dbGaP Accession: [phs000833](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000833.v7.p1).
- Part of [SCAP-T](https://www.scap-t.org/) and originally published in @lake_neuronal_2016.
- Single patient, age = 51, sex = female, NeuN + nuclei from normal mostmortem.
- 6 cortical regions: BA8, BA10, BA21, BA22, BA41, and BA17.
- Annotated into 8 excitatory and inhibitory neuronal subtypes.

This analysis was performed in R except where noted.
The source code and instructions for rerunning the analysis can be found at [github.com/eturkes/lake-2016-snRNAseq](https://github.com/eturkes/lake-2016-snRNAseq).

# Final Results

**Read just this section for the final results of the analysis and a summary of the methods.**

# ~~~ Breakdown of Methods ~~~ {-}

The following top-level sections break down the methods used to perform the analysis and only needs to be read if one is interested.
We start by loading in any required packages and setting some global variables.

```{r}
packages <- c("conflicted", "DT", "SingleCellExperiment", "scater", "ggplot2")
invisible(suppressMessages(lapply(packages, library, character.only = TRUE)))

data_dir <- file.path(getwd(), "..", "data")
results_dir <- file.path(getwd(), "..", "results")
```

# Original Data {.tabset}

This section provides a brief look at the raw data before manipulation.

## Counts

```{r, cache = TRUE}
counts <- read.delim(
  file.path(data_dir, "Lake-2016_Gene_TPM.dat.gz"), "\t", header = FALSE, stringsAsFactors = FALSE)
```

```{r}
datatable(counts[1:5, 1:3])
```

## Clusters

```{r}
clust <- read.table(file.path(data_dir, "Lake-2016_Gene_TPM_Sample-annotation.txt"), header = TRUE)
datatable(clust[1:5, ])
```

# Preliminary Cleaning

Here we provide more informative labels to things, subset to the regions/cell types we are interested in, and transform the data into more convenient formats for downstream analysis.

## Relabeling and Subsetting {.tabset}

A more elegant solution is planned, but for now, the desired regions/cell types can be set here.

```{r}
# TODO: Simplify formatting requirements for the variables.
regions <- "*"
cell_types <- "Ex1|Ex2|Ex3|Ex4|Ex5|Ex6|Ex7|Ex8|In1|In2|In3|In4|In5|In6|In7|In8"
```

### Counts

```{r}
gene_names <- counts[ , 1]
cell_names <- counts[1, ]
cell_names <- as.character(unlist(cell_names))
cell_names <- cell_names[-1]
gene_names <- gene_names[-1]
counts <- counts[-1, -1]
exclude <- duplicated(gene_names)
keep_cells <- cell_names %in% clust[ , 2]
counts <- counts[ , keep_cells]
cell_names <- cell_names[keep_cells]
colnames(counts) <- cell_names
reorder <- order(colnames(counts))
counts <- counts[ , reorder]
datatable(counts[1:5, 1:3])
```

### Clusters

```{r}
clust <- clust[clust[ , 2] %in% colnames(counts), ]
clust <- clust[order(clust[ , 2]), ]
source <- as.character(unlist(clust[ , 4]))
batch <- as.character(unlist(clust[ , 4]))
type <- as.character(unlist(clust[3]))
age <- rep("51yo", times = length(batch))
tmp <- sub("_S.+", "", colnames(counts))
tmp <- matrix(unlist(strsplit(tmp, "_")), ncol = 2, byrow = TRUE)
well <- tmp[ , 2]
plate <- tmp[ , 1]
data <- apply(counts, 1, as.numeric)
data <- t(data)
colnames(data) <- colnames(counts)
data <- data[!exclude, ]
rownames(data) <- gene_names[!exclude]
clust <- data.frame(
  species = rep("homo_sapiens", times = length(type)), cell_type = type, source = source,
  age = age, well = well, batch = batch, plate = plate)
rownames(clust) <- colnames(counts)
datatable(clust[1:5, ])
```

## SingleCellExperiment

```{r}
sce <- SingleCellExperiment(
  assays = list(normcounts = as.matrix(sapply(counts, as.numeric))), colData = clust,
  rowData = gene_names)
sce
```

# QC

These sections assess and correct issues and irregularities in the dataset.

## Normalization

Use the the log2 transformation of the transcript counts.

```{r}
logcounts(sce) <- log2(normcounts(sce) + 1)
rowData(sce)$feature_symbol <- rowData(sce)
```

## Duplicates

Remove features with duplicated names.

```{r}
sce <- sce[!duplicated(rowData(sce)$feature_symbol), ]
```

## ERCC Spike-ins

Identify ERCC spike-ins.

```{r}
isSpike(sce, "ERCC") <- grepl("^ERCC-", rownames(sce))
```

## Library Size

Plot the library size.

```{r}
cell_types <- sort(colnames(colData(sce))[grepl("cell_type", colnames(colData(sce)))])
# for (ct in cell_types) {
#   print(plotColData(sce, aes_string(x = "total_counts", y = "total_features", colour = ct)))}
```

## Detected Genes

Plot the detected genes.

```{r}
# for (ct in cell_types) {
#   print(plotColData(
#     sce, aes_string(x = ct, y = "total_counts", colour = "log10_total_counts")) +
#       theme(axis.text.x = element_text(angle = 90, hjust = 1)))}
```

## Expression

Plot metrics about the expression of features.

```{r}
# plotHighestExprs(sce)
```

```{r}
# plotExprsFreqVsMean(sce)
```

# Dimension Reduction

This section details steps taken to reduce the dimensionality of the dataset through methods like PCA, tSNE, UMAP, etc.

## PCA

We start by performing PCA on all of the genes rather than just HVGs using denoisePCA from the scran package. This function automatically selects PCs by modelling technical noise.

```{r, cache = TRUE}
sce <- denoisePCA(sce, technical = new_trend, approx = TRUE)
```

We plot principal components using `plotPCA` from the `scater` package.

```{r}
for (ct in cell_types) {
  print(plotPCA(sce, colour_by = ct))}
```

# References

This is the concluding section of the document.
Here we write relevant results to disk, output the `sessionInfo`, and create a bibliography for works cited.

```{r}
saveRDS(sce, file = file.path(results_dir, "data", "sce.rds"))

sessionInfo()
```
