---
title: "Lake 2016 snRNAseq Report"
author:
  - name: "Emir Turkes [et2628@cumc.columbia.edu]"
  - name: "Columbia University"
date: '`r strftime(Sys.time(), format = "%B %d, %Y")`'
bibliography: "../lake-2016-snRNAseq.bib"
biblio-style: apalike
link-citations: true
output:
  html_document:
    number_sections: true
    theme: lumen
    highlight: haddock
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
knit:
  (function(inputFile, encoding) {
    rmarkdown::render(
      inputFile, encoding = encoding, output_file = "../results/lake-2016-snRNAseq-report.html")})
---

```{r, include = FALSE}
#    This file is part of lake-2016-snRNAseq.
#    Copyright (C) 2019  Emir Turkes
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

knitr::opts_chunk$set(fig.width = 8.5, fig.height = 7)
```

<style type="text/css">
body {font-size: 16px;}
h1.title {font-size: 35px;}
h1 {font-size: 24px;}
h2 {font-size: 22px;}
h3 {font-size: 20px;}
.toc-content {padding-left: 0px; padding-right: 0px;}
div.tocify {width: 100%;}
.tocify-subheader .tocify-item {font-size: 0.95em; padding-left: 25px; text-indent: 0;}
div.main-container {max-width: none; width: 100%;}
</style>

*This is a broad initial analysis that prepares and characterizes the data for use in other projects.*

The background for this data is as follows:

- dbGaP Accession: [phs000833](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000833.v7.p1).
- Part of [SCAP-T](https://www.scap-t.org/) and originally published in @lake_neuronal_2016.
- Single patient, age = 51, sex = female, NeuN + nuclei from normal mostmortem.
- 6 cortical regions: BA8, BA10, BA21, BA22, BA41, and BA17.
- Annotated into 8 excitatory and inhibitory neuronal subtypes.

This analysis was performed in R except where noted.
The source code and instructions for rerunning the analysis can be found at [github.com/eturkes/lake-2016-snRNAseq](https://github.com/eturkes/lake-2016-snRNAseq).

# Final Results

**Read just the following sub-section for the final results of the analysis and a brief summary of the methods.**

# ~~~ Breakdown of Methods ~~~ {-}

**Sections from here to the end break down the methods used and are optional to read.**

We start by loading in any required packages and setting some global variables.

```{r}
packages <- c("conflicted", "DT", "SingleCellExperiment", "scater", "ggplot2", "BiocFileCache")
invisible(suppressPackageStartupMessages(lapply(packages, library, character.only = TRUE)))

options(stringsAsFactors = FALSE)

data_str <- "Lake-2016_Gene_TPM"
data_dir <- file.path(getwd(), "..", "data")
results_dir <- file.path(getwd(), "..", "results")
```

# Original Data {.tabset}

This section provides a brief look at the raw data before manipulation.

```{r}
bfc <- BiocFileCache(data_dir, ask = FALSE)
counts <- bfcrpath(bfc, file.path(
  "http://genome-tech.ucsd.edu/public/Lake_Science_2016", paste0(data_str, ".dat.gz")))
anno <- bfcrpath(bfc, file.path(
  "http://genome-tech.ucsd.edu/public/Lake_Science_2016", paste0(data_str, "_Sample-annotation.txt")))
rm(bfc)
```

## Counts

```{r}
counts <- read.delim(counts, "\t", header = FALSE)
datatable(counts[1:5, 1:3])
```

## Annotations

```{r}
anno <- read.table(anno, header = TRUE)
datatable(anno[1:5, ])
```

# Preliminary Cleaning

Here we do any data wrangling neccessary to transform the data into more convenient formats for downstream analysis.

### Counts

```{r}
counts <- counts[!duplicated(counts[ , 1]), ]
genes <- counts[ , 1]
genes <- genes[-1]
cells <- as.character(unlist(counts[1, ]))
cells <- cells[-1]
keep <- cells %in% anno[ , 2] # Columns 1 and 2 are not identical.
counts <- counts[-1, -1]
counts <- counts[ , keep]
cells <- cells[keep]
counts <- as.matrix(as.data.frame(lapply(counts, as.numeric)))
colnames(counts) <- cells
counts <- counts[ , order(colnames(counts))] # Counts file is originally unaligned with annotations.
rownames(counts) <- genes
datatable(counts[1:5, 1:3])
```

### Annotations

```{r}
anno <- anno[anno[ , 2] %in% colnames(counts), ]
anno <- anno[order(anno[ , 2]), ]
region <- as.character(unlist(anno[ , 4]))
lake_cluster_name <- as.character(unlist(anno[3]))
tmp <- sub("_S.+", "", colnames(counts))
tmp <- matrix(unlist(strsplit(tmp, "_")), ncol = 2, byrow = TRUE)
well <- tmp[ , 2]
plate <- tmp[ , 1]
anno <- data.frame(
  lake_cluster_name = lake_cluster_name, region = region, well = well, plate = plate)
rownames(anno) <- colnames(counts)
datatable(anno[1:5, ])
```

## SingleCellExperiment

The data is now sufficiently prepared to form a SingleCellExperiment (SCE) object.
Importantly, the data was provided as TPMs not raw counts, so we use the `normcounts` slot and tailor all following analyses to this fact.

```{r}
sce <- SingleCellExperiment(assays = list(normcounts = counts), colData = anno)
rm(counts, anno)
sce
```

# QC

These sections assess and correct issues and irregularities in the dataset.

# References

This is the concluding section of the document.
Here we write relevant results to disk, output the `sessionInfo`, and create a bibliography for works cited.

```{r}
saveRDS(sce, file = file.path(results_dir, "data", "sce.rds"))

sessionInfo()
```
